File: 2ºAÑO/1ºCuatrimestre/PDOO/examen_p1_p2/examen/lib/mazo_sorpresas.rb
# encoding:utf-8

module Civitas
  class Mazo_sorpresas
    
    def initialize (debug = false)  #asi le estas diciendo que si no le pasa nada es a false
      init
      @debug = debug
      @ultima_sorpresa
    end
    
    def init
      @sorpresas = Array.new
      @cartas_especiales = Array.new
      @vacio = false
      @barajada = false
      @usadas = 0
    end
    
    #en ruby no se puede tener dos metodos con el mismo nombre
    
    def Mazo_sorpresas(d)   #dudas con este tambien
      @debug = d
      init
      if @debug == true
        Diario.instance.ocurre_evento("debug ya activado")
      end
    end
    
    def al_mazo(s)
      if @barajada == false
        @sorpresas << s
      end
    end
    
    def ultima_sorpresa
      return @ultima_sorpresa
    end
    
    def siguiente
      if @barajada == false || @usadas == @sorpresas.size
        if !@debug
          @sorpresas.shuffle 
          @usadas = 0
          @barajada = true
        end
      end
      
      @usadas = @usadas + 1
      aux = @sorpresas.at(0)
      @sorpresas.delete_at(0)
      @sorpresas << aux
      @ultima_sorpresa = aux
      return @ultima_sorpresa
    end
    
    def inhabilitar_carta_especial(sorpresa)
      if @sorpresas.include?(sorpresa)
        @sorpresas.delete(sorpresa)
        @cartas_especiales << sorpresa
      end
    end
    
    def habilitar_carta_especial(sorpresa)
      if @cartas_especiales.include?(sorpresa)
        @cartas_especiales.delete(sorpresa)
        @sorpresas << sorpresa
      end
    end
    
    private :init
    
  end
end

File: 2ºAÑO/1ºCuatrimestre/PDOO/Civitas/Civitas/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author Jose
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private Boolean barajada;
    private int usadas;
    private Boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init() {
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(Boolean d) {
        d = debug;
        this.init();
        
        if(debug == true) {
            Diario.getInstance().ocurreEvento("debug ya activado");
        }
    }
    
    MazoSorpresas() {
        this.init();
        debug = false;
    }
    
    void alMazo(Sorpresa s) {
        if(barajada == false) {
            sorpresas.add(s);
        }
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
    
    Sorpresa siguiente() {
        if(barajada == false || usadas == sorpresas.size()) {
            if(debug != true) {
                
                Collections.shuffle(sorpresas);
                usadas = 0;
                barajada = true;
            }
        }
        
        usadas++;
        Sorpresa aux = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(aux);
        ultimaSorpresa = aux;
        
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa) {
        if(sorpresas.contains(sorpresa)) {
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);
            String evento = "inhabilitarCartaEspecial"; 
            Diario.getInstance().ocurreEvento(evento);
            //preguntar funcionamiento diario
        }
        
        
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa) {
        if(cartasEspeciales.contains(sorpresa)) {
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
            String evento = "habilitarCartaEspecial";
            Diario.getInstance().ocurreEvento(evento);
        }
    }
}
